Fix comparisons for complex types using floats in RowContainer

czentgr opened the pull request:
The issue is that `leftValue == rightValue` does not work if both sides are NaN. In fact, none of the comparisons with a NaN work. Equals is false and other comparisons cause an exception.
Therefore, for floating points additional logic is needed to handle the equals case.

The row container implementation contained two bugs when 
using complex types.

1.  Incorrect comparison implementation for floating point 
types when NaN values are used.
2. The comparison flags were not passed down to
lower level functions.
3. Refactor to use SimpleVector::comparePrimitiveAsc in RowContainer 
and ContainerRowSerde for a common comparison function.

In addition, the PR adds testing for floating point using the
RowContainer (and subsequent ContainerRowSerde).

When using floating point values in complex types (ROW, ARRAY, MAP (key))
the following operators are affected by the changes:

Operators:
FilterProject, TopN, TopNRowNumber, OrderBy, MergeExchange,
LocalMerge, HashProbe, NestedLoopJoinProbe

Functions:
array_distinct, array_duplicates, array_except, array_max,
array_min, array_has_duplicates, array_sort, array_sort_desc,
map_union, min, max, min_by, max_by, set_union

The lists may not be complete. Any operator/function that requires a
comparison of elements of complex types that use 
floating point type is affected.

Fixes https://github.com/prestodb/presto/issues/20283

czentgr commented:
Test results prior to code change:

```
czentgr@Christians-MacBook-Pro tests % ./velox_exec_test --gtest_filter=AggregationTest.testFloatingPointArrayNaNGroupBy
Note: Google Test filter = AggregationTest.testFloatingPointArrayNaNGroupBy
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from AggregationTest
[ RUN      ] AggregationTest.testFloatingPointArrayNaNGroupBy
/Users/czentgr/gitspace/velox/velox/exec/tests/utils/QueryAssertions.cpp:1013: Failure
Failed
Expected 1, got 2
1 extra rows, 0 missing rows
1 of extra rows:
        ["NaN",1,2]

0 of missing rows:

Note: DuckDB only supports timestamps of millisecond precision. If this test involves timestamp inputs, please make sure you use the right precision.
DuckDB query: SELECT c0 FROM tmp GROUP BY c0
/Users/czentgr/gitspace/velox/velox/exec/tests/utils/QueryAssertions.cpp:1013: Failure
Failed
Expected 1, got 2
1 extra rows, 0 missing rows
1 of extra rows:
        ["NaN",1,2]

0 of missing rows:

Note: DuckDB only supports timestamps of millisecond precision. If this test involves timestamp inputs, please make sure you use the right precision.
DuckDB query: SELECT c0 FROM tmp GROUP BY c0
[  FAILED  ] AggregationTest.testFloatingPointArrayNaNGroupBy (29 ms)
[----------] 1 test from AggregationTest (29 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (58 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] AggregationTest.testFloatingPointArrayNaNGroupBy

 1 FAILED TEST
```


Test result after fix
```
czentgr@Christians-MacBook-Pro tests % ./velox_exec_test --gtest_filter=AggregationTest.testFloatingPointArrayNaNGroupBy
Note: Google Test filter = AggregationTest.testFloatingPointArrayNaNGroupBy
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from AggregationTest
[ RUN      ] AggregationTest.testFloatingPointArrayNaNGroupBy
[       OK ] AggregationTest.testFloatingPointArrayNaNGroupBy (29 ms)
[----------] 1 test from AggregationTest (29 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (57 ms total)
[  PASSED  ] 1 test.
```

czentgr commented:
@aditi-pandit @mbasmanova if you get a chance please review. This is the fix for the NaN value in array group by issue.

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -123,6 +123,22 @@ class AggregationTest : public OperatorTestBase {
   }
 
+  template <typename T>
+  void testFloatingPointArrayNaNGroupBy() {
```
This change is better tested in a unit test inside velox/exec/tests/ContainerRowSerdeTest.cpp

Also, please, document this behavior in a method comment in .h file.

aditi-pandit replied:
I think https://github.com/facebookincubator/velox/blob/main/velox/exec/tests/RowContainerTest.cpp#L281 was intended to test float comparison. But it didn't go as far as testing when these cases are in Arrays, Complex types. Might be good to enhance the tests here in RowContainer to check for the special floating point values in the Arrays, Maps and Complex types.

mbasmanova commented on velox/exec/ContainerRowSerde.cpp:
```diff
@@ -393,6 +393,44 @@ int compare<TypeKind::VARBINARY>(
+}
+
+template <>
+int compare<TypeKind::DOUBLE>(
```
There is another version of compare that takes 2 ByteStreams. Should it be modified as well?

mbasmanova reviewed (changes requested):
@czentgr Thank you for fixing this. 

Should we also check BaseVector::compare and equalValueAt to see if it has a similar issue?

czentgr commented:
@mbasmanova Thank you for your review. I'm on vacation for the next two weeks and be back Aug 17. Then I will take care of your comments.

aditi-pandit commented:
> @czentgr Thank you for fixing this.
> 
> Should we also check BaseVector::compare and equalValueAt to see if it has a similar issue?

@czentgr : Yes, I too feel that this code change would be needed but the problem has a bigger scope requiring a change at BaseVector::compare and equalValueAt level. So it would be worth spending some time looking at different SQL queries which need a NaN comparison in different places to come up with the most generic solution.

mbasmanova commented:
@czentgr 

> I'm on vacation for the next two weeks and be back Aug 17. Then I will take care of your comments.

Thank you for the heads up.

czentgr commented:
I checked `BaseVector::compare` and `BaseVector::equalValueAt`. Both of these functions are calling a pure virtual `compare` function that is implemented by the more specific `BaseVector` derived classes. For example, I then checked the `SimpleVector` class functions and it handles the `NaN` case (see function `comparePrimitiveAsc` which is called by the implementation of the pure virtual `compare` function of the base class). It appears that for the other vector types `FlatVector`, `ComplexVector` and ConstantVector` `compare` eventually will call `comparePrimitiveAsc` of `SimpleVector`.

The difference to the vector functions and `ContainerRowSerde` is that one side is a vector input which are compared against a byte stream (and not another vector). It doesn't build a vector from the stream itself (which would enable it then to use the vector compare function without running into the problem).

I believe the Vector cases are already handled properly. Perhaps there is something where two streams are compared that could have a similar situation? It doesn't appear that `ContainerRowSerde` implements a compare for two byte streams.

mbasmanova commented:
> It doesn't appear that ContainerRowSerde implements a compare for two byte streams.

I see this method in velox/exec/ContainerRowSerde.h

```
  static int32_t compare(
      ByteStream& left,
      ByteStream& right,
      const Type* type,
      CompareFlags flags);
```

czentgr commented:
Yes. I found this since as well. Which means this piece is also susceptible to the bug. There is lots of the same code with minor differences in how values are accessed from the input (byte stream or vector).

aditi-pandit commented:
> > It doesn't appear that ContainerRowSerde implements a compare for two byte streams.
> 
> I see this method in velox/exec/ContainerRowSerde.h
> 
> ```
>   static int32_t compare(
>       ByteStream& left,
>       ByteStream& right,
>       const Type* type,
>       CompareFlags flags);
> ```

@czentgr : In general RowContainer::compare (which is used in HashAgg, HashJoin, Window etc) would be the origin of these calls.
There is a variant being used from 
https://github.com/facebookincubator/velox/blob/main/velox/exec/RowContainer.cpp#L521
that invokes the compare on 2 ByteStreams. We should fix this.

aditi-pandit commented:
> Yes. I found this since as well. Which means this piece is also susceptible to the bug. There is lots of the same code with minor differences in how values are accessed from the input (byte stream or vector).

We should standardize on this style https://github.com/facebookincubator/velox/blob/main/velox/exec/RowContainer.h#L1023

czentgr commented:
> > Yes. I found this since as well. Which means this piece is also susceptible to the bug. There is lots of the same code with minor differences in how values are accessed from the input (byte stream or vector).
> 
> We should standardize on this style I feel https://github.com/facebookincubator/velox/blob/main/velox/exec/RowContainer.h#L1023

Yes, this should be a better fix compared to making new templated functions. There is some addition with regards to the (possibly requested) ordering but lets see if this can be worked out.

aditi-pandit commented on velox/exec/ContainerRowSerde.cpp:
```diff
@@ -587,7 +600,20 @@ int32_t compare(
   T leftValue = left.read<T>();
   T rightValue = right.read<T>();
-  auto result = leftValue == rightValue ? 0 : leftValue < rightValue ? -1 : 1;
+  int32_t result{0};
```
This code looks exactly the same as for the other function. Can we abstract this in a common function ?

aditi-pandit commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +144,36 @@ TEST_F(ContainerRowSerdeTest, nested) {
   testRoundTrip(nestedArray);
 }
 
+TEST_F(ContainerRowSerdeTest, compareDoubleEquals) {
```
In general it seems like all the comparison function tests are done through RowContainerTests. That is the main caller and touch-point for Velox operators. Please can you add tests there also.

czentgr replied:
I will take a look at those tests.

aditi-pandit commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +144,36 @@ TEST_F(ContainerRowSerdeTest, nested) {
+  auto leftPosition = serialize(data);
+  HashStringAllocator::prepareRead(leftPosition.header, left);
+
+  SelectivityVector allRows(data->size());
```
Do we need this DecodedVector ? Can we not compare with a FlatVector ?

czentgr replied:
Yes. The API requires a DecodedVector (there is a difference in the interface and implementation). When using FlatVector (which I originally tried) compilation fails with
```
/Users/czentgr/gitspace/velox/velox/exec/tests/ContainerRowSerdeTest.cpp:160:18: error: no matching function for call to 'compare'
    EXPECT_EQ(0, ContainerRowSerde::compare(left, data, i, flags));
...
/Users/czentgr/gitspace/velox/./velox/exec/ContainerRowSerde.h:35:18: note: candidate function not viable: no known conversion from 'FlatVectorPtr<EvalType<double>>' (aka 'shared_ptr<FlatVector<double>>') to 'const facebook::velox::DecodedVector' for 2nd argument
  static int32_t compare(
```

aditi-pandit commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +144,36 @@ TEST_F(ContainerRowSerdeTest, nested) {
+  HashStringAllocator::prepareRead(rightPosition.header, right);
+  auto type = data->type();
+
+  for (auto i = 0; i < data->size(); i++) {
```
Try different combinations where Nan is compared with non-Nan values also as those code paths have changed now.

Also add tests for ascending/descending variation of flags.

czentgr commented:
@mbasmanova @aditi-pandit Please take another look when you get a chance. I've added comprehensive tests for the complex types based on the existing `RowContainer` test. These tests test both types of interfaces - 1. two byte streams and 2. byte stream and decoded vector .
As a result, I found a second bug where the compare flags were not passed down to the lower comparison functions for the complex types which is fixed in this PR as well.

mbasmanova commented:
CC: @duanmeng 

@czentgr Does this logic match both Presto and Spark semantics? Since this is shared code we need to make sure it works for both.

mbasmanova commented on velox/exec/ContainerRowSerde.cpp:
```diff
@@ -332,6 +332,28 @@ int compareSwitch(
     CompareFlags flags);
 
+template <typename T>
+int32_t compareNative(
```
Should we move the tests to velox/exec/tests/ContainerRowSerdeTest.cpp ?

It would be nice to update the comments for the ContainerRowSerde API to clarify the semantics of comparing floating point values.

czentgr replied:
We could. I put this in the RowContainer because the test is at that level and it contains the simple type tests - so they are all together.
In the RowContainer there is a function that operates on the vector level (as the row must not be serialized) which is called `comparePrimitiveAsc`. Perhaps `native` isn't quite the right word for it and I should use `comparePrimitive`?

mbasmanova replied:
Is this a copy of comparePrimitiveAsc from SimpleVector.h? If so, perhaps, modify SimpleVector.h to make comparePrimitiveAsc a public static method and then re-use here.

czentgr replied:
No. It is a private static function in `RowContainer` (see RowContainer.h). This function also always compares in ascending order (and the caller takes care of switching the result based on the compare flag) while the new function allows for also descending order results by passing the compare flags. But otherwise, they are the same. I can see that there should be some synergy here.
Interesting, the RowContainer version is a copy of the SimpleVector one.

mbasmanova replied:
It would be nice to refactor to ensure there is only one copy of this logic.

czentgr replied:
I've refactored the code to use the SimpleVector implementation.

As for the comments of the ContainerRowSerde API. Isn't this more generic around data type value handling for floating point, similar to how Spark has it documented?
Maybe something for develop/types.rst?

mbasmanova replied:
> I've refactored the code to use the SimpleVector implementation.

Thanks.

> As for the comments of the ContainerRowSerde API. Isn't this more generic around data type value handling for floating point, similar to how Spark has it documented?
> Maybe something for develop/types.rst?

It would be nice to have documentation both in the code and on the website.

czentgr commented:
> CC: @duanmeng
> 
> @czentgr Does this logic match both Presto and Spark semantics? Since this is shared code we need to make sure it works for both.

@mbasmanova I believe this to be the case. In the end we didn't invent new semantics for handling floating point NaN values. We applied the existing logic to compare complex type elements when resolved to the native floating point type that is encapsulated in the complex type. 
The handling of NaN in Spark is 

> NaN Semantics
> 
> There is special handling for not-a-number (NaN) when dealing with float or double types that do not exactly match standard floating point semantics. Specifically:
> 
>     - NaN = NaN returns true.
>     - In aggregations, all NaN values are grouped together.
>     - NaN is treated as a normal value in join keys.
>     - NaN values go last when in ascending order, larger than any other numeric value.
> 
from https://spark.apache.org/docs/latest/sql-ref-datatypes.html
This matches the implementation for the comparison added here.

As for Presto, the documentation on this topic isn't as exhaustive as Spark's. I checked and did not find a similar detailed explanation for the behavior regarding the special floating point values.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -278,7 +279,44 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+        sorted.push_back(decoded.valueAt<T>(irow.first));
+      }
+    }
+    return vectorMaker.template flatVectorNullable<T>(sorted);
```
An instance of VectorMaker is available in the base class (vectorMaker_). Also, there is a method makeNullableFlatVector<T>(sorted) available in the base class as well.

czentgr replied:
Right, left over from the previous test. I didn't touch much of that code other than trying to refactor obvious common code.

mbasmanova replied:
It would be nice to clean that up. Any chance you can send a refactoring PR?

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -278,7 +279,44 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+
+  template <typename T>
+  VectorPtr createResultVectorForSimpleComparison(
+      const std::vector<std::pair<int, char*>>& indexedRows,
```
int -> vector_size_t

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -278,7 +279,44 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
   template <typename T>
-  void testCompareFloats(TypePtr type, bool ascending, bool nullsFirst) {
+  void setupFloatTestExpectedOutput(
+      T& initialOrder,
```
Is T always std::vector<std::optional<U>>? If so, maybe it would be clearer as

```
  template <typename T>
  void setupFloatTestExpectedOutput(
      std::vector<std::optional<T>>& initialOrder,
```

It is not clear what this method does. Maybe add a comment to clarify?

czentgr replied:
I added a comment. 

> Is T always std::vector<std::optional>?

No, they are not. The initial order type changes based on the complex type. In the end these are row, map or array vectors and not only `std::vector<std::optional<T>>` (see the various types for `expectedOrder` in the test functions.
But each time the reordering of the elements based on the test inputs is the same, hence I made a templated version. The exception is the array test. But otherwise, the function is shared for the simple, map and row test. Admittedly, the simple and row test have the same type while map is different.

mbasmanova reviewed (commented):
@czentgr The code changes look good. I find it hard to read and reason about the tests though. I wonder if there is a way to make these clearer.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,8 +278,50 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+  /// vector elements and, if needed, reordered.
   template <typename T>
-  void testCompareFloats(TypePtr type, bool ascending, bool nullsFirst) {
+  void setupFloatTestExpectedOutput(
```
Looking at this method some more, it seems like it takes a list of values sorted in ascending order with nulls first and re-orders the list according to specified ascending and nullsFirst flags. If that's the case, I suggest to rename this method to countLeadingNulls and implement it like this:

```
  // Count number of nulls at the start of the list.
  template <typename T>
  static size_t countLeadingNulls(std::vector<std::optional<T>>& values) {
    int32_t numNulls = 0;
    for (auto i = 0; i < values.size(); ++i) {
      if (!values[i].has_value()) {
        ++numNulls;
      } else {
        break;
      }
    }

    return numNulls;
  }

  // Given a list of values sorted in ascending order with nulls first re-orders
  // the elements to reverse the order and / or pu nulls last.
  template <typename T>
  void changeSortingOrder(
      std::vector<std::optional<T>>& values,
      bool ascending,
      bool nullsFirst) {
    if (ascending) {
      if (nullsFirst) {
        // Nothing to do. 'values' are expected to be already sorted in
        // ascending order with nulls first.
      } else {
        // Move the nulls from the start of the list to the end of the list.
        const auto numNulls = countLeadingNulls(values);
        for (auto i = 0; i < values.size() - numNulls; ++i) {
          values[i] = values[i + numNulls];
        }

        for (auto i = values.size() - numNulls; i < values.size(); ++i) {
          values[i].reset();
        }
      }
    } else {
      if (nullsFirst) {
        // Keeps nulls at the start of the list. Reverse the order of remaining
        // non-null elements.
        const auto numNulls = countLeadingNulls(values);
        std::reverse(values.begin() + numNulls, values.end());
      } else {
        // Reverse the order of all elements.
        std::reverse(values.begin(), values.end());
      }
    }
  }
```

Then, it can be called like this in all 3 places.

```
changeSortingOrder(expectedOrder, ascending, nullsFirst);
```

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -304,23 +346,147 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+
+    // Verify compare method with row and decoded vector as input
+    {
+      std::sort(
```
This usage of std::sort is very strange since comparison function usually compares two entries from the collector being sorted, but here we compare an element from a RowContainer with a value from a DecodedVector. This happens to work that Each row in a RowContainer has an equivalent row in a DecodedVector, but it is confusing to reader.

Let's find a different, less confusing, way to test comparison between RowContainter row and DecodedVector row.

czentgr replied:
Right. The idea here was - with the assumption that decodedVector and the rows have the same order - to let the `sort` function do the comparison to test all value combinations.
I see how the initial author used this to write a short test that tests the specific compare API.
Regarding a replacement we could set up two input vectors (one for the row container, one for the decoded vector) and call the compare API. However, as expected output the result matrix would need to be generated based on the vector contents which includes implementing a function to do the value comparisons properly.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -304,23 +346,147 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+      assertEqualVectors(expected, result);
+    }
+
+    // Verify compareRows method with row as input.
```
Is this different from the first test case (// Verify compare method with two rows as input) ? These seem identical.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -304,23 +346,147 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+      assertEqualVectors(expected, result);
+    }
+
+    // Verify compareRows method with default compare flags.
```
What's the purpose of this test? Shouldn't the caller call this method 4 times with all combinations of ascending and nullsFirst? This test seems identical to the first one (// Verify compare method with two rows as input).

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -304,23 +346,147 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+         {{min}}});
+    int numRows = values->size();
+
+    SelectivityVector allRows(numRows);
```
The logic for storing all rows of a vector in a RowContainer repeats multiple times. Let's extract a helper method to reduce copy-paste.

```
  // A list of rows with unique zero-based indices.
  using IndexedRows = std::vector<std::pair<vector_size_t, char*>>;

  static std::vector<char*> store(
      DecodedVector& decoded,
      vector_size_t numRows,
      RowContainer& rowContainer) {
    std::vector<char*> rows(numRows);
    for (size_t i = 0; i < numRows; ++i) {
      rows[i] = rowContainer.newRow();
      rowContainer.store(decoded, i, rows[i], 0);
    }

    return rows;
  }

  static IndexedRows indexRows(const std::vector<char*>& rows) {
    IndexedRows indexedRows;
    indexedRows.reserve(rows.size());
    for (auto i = 0; i < rows.size(); ++i) {
      indexedRows.push_back({i, rows[i]});
    }
    return indexedRows;
  }


    int numRows = values->size();
    DecodedVector decoded(*values);
    auto rows = store(decoded, numRows, *rowContainer);
    auto indexedRows = indexRows(rows);
```

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -304,23 +346,147 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
       if (!nullsFirst) {
-        for (int i = 0; i < expectedOrder.size() - 1; ++i) {
-          expectedOrder[i] = expectedOrder[i + 1];
+        for (int i = 0; i < expectedOrder.size() - 2; ++i) {
```
We should be able to re-use changeSortingOrder function here.

```
changeSortingOrder(expectedOrder, ascending, nullsFirst);
```

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -363,17 +522,130 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+  }
+
+  template <typename T>
+  VectorPtr createResultVectorForMapComparison(
```
This template method, createResultVectorForArrayComparison and createResultVectorForSimpleComparison can be replaced with a simpler non-template method:

```
  VectorPtr copyValues(
      const IndexedRows& indexedRows,
      const VectorPtr& values) {
    const auto numRows = indexedRows.size();
    auto copy = BaseVector::create(values->type(), numRows, pool());
    for (auto i = 0; i < numRows; ++i) {
      copy->copy(values.get(), i, indexedRows[i].first, 1);
    }
    return copy;
  }

      auto result = copyValues(indexedRows, values);
      assertEqualVectors(expected, result);
`

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -937,24 +1302,80 @@ TEST_F(RowContainerTest, alignment) {
   // Verify ascending order
-  testCompareFloats<float>(REAL(), true, true);
-  testCompareFloats<float>(REAL(), true, false);
+  testCompareSimpleFloats<float>(REAL(), true, true);
```
There is a lot of copy-paste here. Consider introducing a helper function:

```
  template <typename T>
  void testCompareSimpleFloats(const TypePtr& type) {
    // Verify ascending order.
    testCompareSimpleFloats<float>(REAL(), true, true);
    testCompareSimpleFloats<float>(REAL(), true, false);
    // Verify descending order.
    testCompareSimpleFloats<float>(REAL(), false, true);
    testCompareSimpleFloats<float>(REAL(), false, false);
  }

TEST_F(RowContainerTest, compareFloat) {
  testCompareSimpleFloats<float>(REAL());
}

TEST_F(RowContainerTest, compareDouble) {
  testCompareSimpleFloats<double>(DOUBLE());
}
```

Also, add SCOPED_TRACE to allow for easier troubleshooting of test failures.

```
  template <typename T>
  void testCompareSimpleFloats(TypePtr type, bool ascending, bool nullsFirst) {
    SCOPED_TRACE(fmt::format("{}, ascending = {}, nullsFirst = {}", type->toString(), ascending, nullsFirst));
```

Same for arrays, maps and structs.

mbasmanova reviewed (commented):
@czentgr I took a closer look at the test and made a number of suggestions on how to improve. You can also see the suggested modifications in "Review comments" commit in https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:fix_nan_compare_array?expand=1

aditi-pandit commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -937,24 +1302,80 @@ TEST_F(RowContainerTest, alignment) {
+}
+
+TEST_F(RowContainerTest, compareRowTypeDouble) {
+  static const std::string colName{"floatingPoint"};
```
Change this name to "doubleCol"

aditi-pandit commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -363,17 +522,130 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+        expectedOrder, ascending, nullsFirst);
+    VectorPtr expected = makeNullableMapVector<T, int32_t>(expectedOrder);
+
+    // Verify compare method with two rows as input
```
It seems to me that the tests at an API level are similar across Simple, Array, Map and Row. So a common function can be refactored

(My code is at a semantic level and you will have to figure exact syntax)
```
TypePtr t could be REAL(), ARRAY(REAL()), MAP(REAL()), ROW(REAL())
void testRowContainerCompareAPIs(TypePtr t, ....) {
auto createResult(TypePtr t...) -> VectorPtr {
if (t->isPrimitiveType()) {
  return createResultVectorForSimpleComparison<T>(indexedRows, decoded);
} else (t->isArray()) {
  return createResultVectorForArrayComparison<T>(indexedRows, decoded);
} else (t->isMap()) {
  return createResultVectorForMapComparison<T>(...)
}...
};
{
      std::sort(rows.begin(), rows.end(), [&](const char* l, const char* r) {
        return rowContainer->compare(l, r, 0, {nullsFirst, ascending}) < 0;
      });
      VectorPtr result = BaseVector::create(type, numRows, pool_.get());
      rowContainer->extractColumn(rows.data(), numRows, 0, result);
      assertEqualVectors(expected, result);
    }

    // Verify compare method with row and decoded vector as input
    {
      std::sort(
          indexedRows.begin(),
          indexedRows.end(),
          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
            return rowContainer->compare(
                       l.second,
                       rowContainer->columnAt(0),
                       decoded,
                       r.first,
                       {nullsFirst, ascending}) < 0;
          });
      auto result = createResult(type, indexedRows, decoded);
      assertEqualVectors(expected, result);
    }

    // Verify compareRows method with row as input.
    {
      std::sort(
          indexedRows.begin(),
          indexedRows.end(),
          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
            return rowContainer->compareRows(
                       l.second, r.second, {{nullsFirst, ascending}}) < 0;
          });
      auto result = createResult<T>(indexedRows, decoded);
      assertEqualVectors(expected, result);
    }

    // Verify compareRows method with default compare flags.
    if (ascending && nullsFirst) {
      std::sort(
          indexedRows.begin(),
          indexedRows.end(),
          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
            return rowContainer->compareRows(l.second, r.second) < 0;
          });
      auto result =  createResult<T>(indexedRows, decoded);
      assertEqualVectors(expected, result);
    }
```

That would reduce the code quite a bit and make it easier to follow.

czentgr replied:
I've refactored the code more to have a better separation between the etup for each type and the actual tests that are run.

czentgr reviewed (commented):
@mbasmanova Thank you for your in-depth review and proposed changes. I've applied your proposed changes and made additional changes to reduce duplication further (separation of setup and test - base don Aditi's comment).
I was playing with a matrix comparison test for the compare of decoded vector and row by not using sort. But I think there is some elegance to the currently used method.

mbasmanova commented:
> I was playing with a matrix comparison test for the compare of decoded vector and row by not using sort. But I think there is some elegance to the currently used method.

That's fair. Maybe add a comment in the code to explain the usage of std::sort.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+    auto rows = storeRows(decoded, numRows, *rowContainer);
+    auto indexedRows = indexRows(rows);
+
+    std::function<int(const char*, const char*)> sortRows = [&](const char* l,
```
sortRows -> compare since this lambda compares values, not sorts them

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
-      VectorPtr result = vectorMaker.template flatVectorNullable<T>(sorted);
-      assertEqualVectors(expected, result);
+    std::function<int(const std::pair<int, char*>, const std::pair<int, char*>)>
+        sortDecodedVectorAndRows = [&](const std::pair<int, char*>& l,
```
sortDecodedVectorAndRows -> compareDecoded or similar

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+      const TypePtr& type,
+      const VectorPtr& values,
+      const VectorPtr& expected,
+      bool ascending,
```
maybe pass std::optional<CompareFlags>; this will make the code a bit easier to read

czentgr replied:
If I'm understanding correctly, then it means we would skip passing in ascending/nullsFirst for the default values? Not sure what the purpose of using optional would be.

mbasmanova replied:
With std::optional<CompareFlags> compareFlags argument we would pass compareFlags to the API based on whether the flags are set or not. I think this will make code cleaner and more readable.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+
+  template <typename T>
+  void
+  testComparePrimitiveFloats(TypePtr type, bool ascending, bool nullsFirst) {
```
pass type as const &

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+    auto nan = std::numeric_limits<T>::quiet_NaN();
+    auto inf = std::numeric_limits<T>::infinity();
+
+    VectorPtr values = makeNullableFlatVector<T>(
```
auto

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+
+    std::vector<std::optional<T>> expectedOrder = {
+        std::nullopt, lowest, 0.0, min, max, inf, nan};
+    changeSortingOrder<T>(expectedOrder, ascending, nullsFirst);
```
Maybe rearrange the code to avoid re-creating values and expectedOrder for each combination of compare flags. E.g. remove the flags from the signature and add a loop here.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+    VectorPtr values = makeNullableFlatVector<T>(
+        {std::nullopt, max, inf, 0.0, nan, lowest, min});
+
+    std::vector<std::optional<T>> expectedOrder = {
```
expectedOrder name is confusing; if it is expected, then why changing it?

expectedOrder -> ascNullsFirstOrder or defaultOrder

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+  template <typename T>
+  void testCompareArrayFloats(TypePtr type, bool ascending, bool nullsFirst) {
+    // NOTE: set 'isJoinBuild' to true to enable nullable sort key in test.
+    auto lowest = std::numeric_limits<T>::lowest();
```
this block of code repeats...

mbasmanova reviewed (commented):
@czentgr Thank you for iterating on this PR. The test looks much cleaner. Looks great % a few minor comments.

aditi-pandit commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+      if (!values[i].has_value()) {
+        ++numNulls;
+      } else {
+        break;
```
This might be more readable as a while loop

```
while(!values[i].has_value()) { 
  ++numNulls;
 ++i;
 }
return numNulls;
```

aditi-pandit commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+  }
+
+  static std::vector<char*> storeRows(
+      DecodedVector& decoded,
```
decoded is const& ?

aditi-pandit commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+  }
+
+  // A list of rows with unique zero-based indices.
+  using IndexedRows = std::vector<std::pair<vector_size_t, char*>>;
```
Nit : Rename std::pair<vector_size_t, char*>> to RowAndIndex and the vector of it to RowAndIndexVector ?

czentgr replied:
I used `IndexAndRowVector` because of the way the pair is defined (first, second).

aditi-pandit commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+    return rows;
+  }
+
+  static IndexedRows indexRows(const std::vector<char*>& rows) {
```
Move this just below the typedef for the IndexedRows type so that its clear how it is built.

aditi-pandit reviewed (commented):
Thanks @czentgr. This is looking good.

aditi-pandit commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,304 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
-      rows[i] = rowContainer->newRow();
-      rowContainer->store(decoded, i, rows[i], 0);
-      indexedRows[i] = std::make_pair(i, rows[i]);
+  static size_t countLeadingNulls(const std::vector<std::optional<T>>& values) {
```
Does this need to be a static function ? Can it be made non-static ?

aditi-pandit replied:
The return type can be vector_size_t to be consistent with the rest of the code.

aditi-pandit commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,304 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+  // A list of rows with unique zero-based indices.
+  using IndexAndRowVector = std::vector<IndexAndRow>;
+
+  static IndexAndRowVector indexRows(const std::vector<char*>& rows) {
```
Does this function need to be static ? From the usage seems that is not required.

aditi-pandit commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -277,121 +278,304 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+    return copy;
+  }
+
+  static std::vector<char*> storeRows(
```
Same comment here about static.

aditi-pandit reviewed (approved):
Thanks @czentgr ! Have few nits.

aditi-pandit commented:
@czentgr : Please can you rebase your code and push again to retry the linux-presto-fuzzer-run. The log doesn't clearly indicate the problem.

czentgr commented:
I created a new issue (https://github.com/facebookincubator/velox/issues/6757) for the fuzzer failure. I was able to repro the fuzzer problem on the main code without the changes in this PR.

czentgr commented:
@mbasmanova do you think the current changes are ok?
The latest failures are unrelated to this change
1. linux-build failure:  all unit tests passed but the memory xml doc for one test appears to not have the correct structure and upload failed.
2. linux-presto-fuzzer-run: the fuzzer found a possible problem comparing an aggregation on a timestamp column that should cause overflow - but for one plan does not (I'm looking into that one to see how this all works). Will open an issue if I can't find an matching existing one.

mbasmanova commented:
@czentgr What's the status of this PR. I see that CI is red.

czentgr commented:
@mbasmanova It has been plagued by pipeline issues. Some failing tests have been addressed. There are two failures
1. MultiFragmentTest.maxBytes failed. First time this test failed (having run many pipelines since this code has been rebased multiple times). Also @aditi-pandit saw this once so likely unrelated. However, I tried to repro the problem but after 300+ iterations was not able to repro it. I opened an issue that it may be flaky.
2. A problem with the xml document generated by the velox_memory_test. When uploading the results it claims that 
```
failed uploading test results: File: tmp/test_xml_output/velox_memory_test_72.xml had the following problems:
  * invalid top level end element: testsuite
  * invalid top level end element: testsuites
  * invalid top level end element: testsuites
 ```
Other PRs have the same issue except there it is different memory test number.

I think I addressed all the review comments. I'm rebasing again and hoping the pipeline will pass.

mbasmanova commented:
@czentgr Would you file GitHub issue for test failure with details? We need to look into these and fix.

czentgr commented:
@mbasmanova Yes, for the MultiFragmentTest.maxBytes failure I have opened https://github.com/facebookincubator/velox/issues/6984.

For the uploading of the artifacts I can open an issue. It seems to affect PRs that have a test failure (like this PR had). Another example of this is pipeline https://app.circleci.com/pipelines/github/facebookincubator/velox/35382/workflows/f0bbb571-b29d-44a0-b23c-bc5022f056e4/jobs/233104
I will open an issue for this as well.

czentgr commented:
@mbasmanova this time build and tests look better. This time there is a fuzzer failure. I think it is related to the switch expression for which an issue exists already (though it mentions array_union which contains a switch).

mbasmanova reviewed (commented):
Let's document the rules for comparing floating point types in https://facebookincubator.github.io/velox/develop/types.html

mbasmanova reviewed (commented):
What functions / operators are affected by this change? OrderBy, TopN, Join? Anything else? Let's describe the issue in terms of user-visible impact in the PR description.

czentgr commented:
@mbasmanova 

> Let's document the rules for comparing floating point types in https://facebookincubator.github.io/velox/develop/types.html
Yes, I saw you had opened issue https://github.com/facebookincubator/velox/issues/6717. I suppose we do this in a different PR?

> What functions / operators are affected by this change? OrderBy, TopN, Join?
Basically, anywhere where comparison of complex types is required. The comparison is now equivalent to how scalar types are compared (the complex types are serialized into their scalar type elements and compared). From my understanding there was an assumption that this was intended originally.

I will try to come up with a list to add to the description.

czentgr commented:
@mbasmanova do you think this is good now? The linux-build fails due a test around memory arbitration not passing. There is an issue open for these flaky tests.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -284,121 +285,304 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+  static size_t countLeadingNulls(const std::vector<std::optional<T>>& values) {
+    int32_t numNulls{0};
+    int32_t i{0};
+    while (i < values.size() && !values[i].has_value()) {
```
nit: perhaps, this can be simplified using a for loop

```
for (auto i = 0 ; i < values.size(); ++i) {
   if (values[i].has_value()) {
      return i;
   }
}

return values.size();
```

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -284,121 +285,304 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
-        std::reverse(expectedOrder.begin() + 1, expectedOrder.end());
+  // Given a list of values sorted in ascending order with nulls first re-orders
+  // the elements to reverse the order and / or put nulls last.
+  // Returns a reordered list.
```
This comment is confusing since this is a void function, i.e. it doesn't return anything.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -284,121 +285,304 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+  // the elements to reverse the order and / or put nulls last.
+  // Returns a reordered list.
+  template <typename T>
+  void changeSortingOrder(
```
Thank you for re-writing this function. The new logic is much clearer.

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -284,121 +285,304 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+        flags.has_value() ? flags.value().ascending : true,
+        flags.has_value() ? flags.value().nullsFirst : true));
+
+    // NOTE: set 'isJoinBuild' to true to enable nullable sort key in test.
```
Drop "NOTE:"

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -284,121 +285,304 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+
+  template <typename T>
+  void testCompareArrayFloats(const TypePtr& type) {
+    // NOTE: set 'isJoinBuild' to true to enable nullable sort key in test.
```
What does this note applies to? It is not clear.

czentgr replied:
This comment was present in the previous code that was refactored.

czentgr replied:
Upon closer, look, the `RowContainer` has an `isNullableKeys` option and the tests turns that off (using a variable called `isJoinBuild` in the wrapper function). Which is interesting because the did have null values in the input.

mbasmanova reviewed (approved):
@czentgr Looks good % a few nits.

mbasmanova commented:
@czentgr Seeing lots of lint warnings like this and CI is red. Please, take a look.

<img width="628" alt="Screenshot 2023-10-23 at 9 05 38 AM" src="https://github.com/facebookincubator/velox/assets/27965151/c112d8a9-f7b2-4942-80fd-26106eb7f000">

czentgr commented:
There is one more linter warning. What is it about? Should I address it?

mbasmanova commented on velox/exec/tests/RowContainerTest.cpp:
```diff
@@ -284,121 +285,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
-      }
-      VectorPtr result = vectorMaker.template flatVectorNullable<T>(sorted);
-      assertEqualVectors(expected, result);
+    std::function<int(const std::pair<int, char*>, const std::pair<int, char*>)>
```
<img width="565" alt="Screenshot 2023-10-23 at 1 07 15 PM" src="https://github.com/facebookincubator/velox/assets/27965151/06f1fa65-89df-4e12-9325-92a34e3f4750">
